distsys/archetype_resources/tcp_mailboxes.go

```diff
@@ -0,0 +1,331 @@
+							// could be done by:
+							// - send a unique incrementing counter on pre-commit
+							// - commit-er sends this id on commit, and we store this commit attempt in a persistent set
+							//   alongside the buffer itself
+							// - receiving two commits for the same id can be detected by looking in the set, in which
+							//   case do nothing and reply that the thing is done.
+							// - starting a new critical section sends the id of the remote's last commit, which
+							//   indicates that commit was successful, and that the id can be removed from the set
+							if !hasBegun {
+								err = encoder.Encode(true)
+								if err != nil {
+									continue
+								}
+							} else {
+								// FIXME: this is weak to restarts, but fixing that without proper context is really hard
+								// at least, in this case the msgChannel will function as a rate limiter, so
+								// crash-free operation shouldn't do anything weird
+								for _, elem := range localBuffer {
+									msgChannel <- elem
+								}
+								err = encoder.Encode(false)
+								if err != nil {
+									continue
+								}
+								hasBegun = false
+							}
+						}
+					}
+					err = conn.Close()
+					if err != nil {
+						log.Printf("error closing connection: %s", err.Error())
+					}
+				}()
+			}
+		}()
+
+		return &tcpMailboxesLocalArchetypeResource{
+			msgChannel: msgChannel,
+		}
+	})
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) Abort() chan struct{} {
+	res.readBacklog = append(res.readBacklog, res.readsInProgress...)
+	res.readsInProgress = nil
+	return nil
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) PreCommit() chan error {
+	return nil
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) Commit() chan struct{} {
+	res.readsInProgress = nil
+	return nil
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) ReadValue() (distsys.TLAValue, error) {
+	// if a critical section previously aborted, already-read values will be here
+	if len(res.readBacklog) > 0 {
+		value := res.readBacklog[0]
+		res.readBacklog[0] = distsys.TLAValue{}
+		res.readBacklog = res.readBacklog[1:]
+		return value, nil
+	}
+
+	// otherwise, either pull a notification + atomically read a value from the buffer, or time out
+	select {
+	case msg := <-res.msgChannel:
+		res.readsInProgress = append(res.readsInProgress, msg)
+		return msg, nil
+	case <-time.After(20 * time.Millisecond):
+		return distsys.TLAValue{}, distsys.ErrCriticalSectionAborted
+	}
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) WriteValue(value distsys.TLAValue) error {
+	panic(fmt.Errorf("attempted to write value %v to a local mailbox archetype resource", value))
+}
+
+type tcpMailboxesRemoteArchetypeResource struct {
+	distsys.ArchetypeResourceLeafMixin
+	dialAddr string
+
+	inCriticalSection bool
+	conn              net.Conn
+	connEncoder       *gob.Encoder
+	connDecoder       *gob.Decoder
+	sendBuffer        []distsys.TLAValue
```

Comment:
Why do we never used the `sendBuffer`?

Reply:
We do use it... kind of, but it's never written. Probably better to remove and replace code path with a panic; it's part of a not-supported-yet recovery mechanism.